### PR TITLE
feat(fake/mock): allow pre-programming methods with callbacks

### LINKED ISSUE #171: feat: set arg expectations on methods
## Summary

What:

Test double methods that can set up to return values should have arg expectations defined as well.

Why:

It's common to set up a method like "expect this method to take in this arg and return this value"

## Acceptance Criteria

Below is a list of tasks that must be completed before this issue can be closed.

- [ ] Write documentation
- [ ] Write unit tests
- [ ] Write integration tests

## Example Pseudo Code (for implementation)

```typescript
fake.method("blah").withArgs("hello").willReturn("world");
```

### LINKED ISSUE #173: feat(fake/mock): allow pre-programming methods with callbacks
## Summary

What:

Allow the following ...

```typescript
describe("...", () => {
  it("...", () => {
    const mockFiveService = Mock(TestObjectFiveServiceMultipleArgs)
      .create();

    const mockFive = Mock(TestObjectFiveMultipleArgs)
      .withConstructorArgs(mockFiveService)
      .create();

    assertEquals(mockFive.is_mock, true);
    assertEquals(mockFiveService.is_mock, true);

    mockFiveService
      .method("get")
      .willReturn((key: string, defaultValue: number | string) => {
        if (key == "host" && defaultValue == "localhost") {
          return null;
        }

        if (key == "port" && defaultValue == 5000) {
          return 4000;
        }

        return undefined;
      });

    // `false` because `mockFiveService.get("port") == 3000`
    assertEquals(mockFive.send(), false);

    mockFiveService
      .method("get")
      .willReturn((key: string, defaultValue: string | number) => {
        if (key == "host" && defaultValue == "localhost") {
          return "locaaaaaal";
        }

        if (key == "port" && defaultValue == 5000) {
          return 4000;
        }

        return undefined;
      });

    // `true` because `mockFiveService.get("port") != 3000`
    assertEquals(mockFive.send(), true);
  });
});


class TestObjectFiveMultipleArgs {
  private readonly service: TestObjectFiveServiceMultipleArgs;

  public constructor(service: TestObjectFiveServiceMultipleArgs) {
    this.service = service;
  }

  public send(): boolean {
    const host = this.service.get<string>("host", "localhost");
    const port = this.service.get<number>("port", 5000);

    if (host == null) {
      return false;
    }

    if (port === 3000) {
      return false;
    }

    return true;
  }
}

class TestObjectFiveServiceMultipleArgs {
  #map = new Map<string, unknown>();
  public get<T>(key: string, defaultValue: T): T {
    return this.#map.get(key) as T ?? defaultValue;
  }
}

```

Why:

Screen Shot 2022-12-11 at 00 48 42

## Acceptance Criteria

Below is a list of tasks that must be completed before this issue can be closed.

- [x] Write documentation: https://github.com/drashland/website-v2/pull/193
- [x] Write unit tests
- [x] Write integration tests

## Example Pseudo Code (for implementation)

```typescript
// Add example pseudo code for implementation
```

### PR DESCRIPTION
**Pull request by crookse**
Closes #173 

## Docs

- [Fakes](https://drash-land-git-rhum-will-return-callback-drash-land.vercel.app/rhum/v2.x/tutorials/test-doubles/fakes)
- [Mocks](https://drash-land-git-rhum-will-return-callback-drash-land.vercel.app/rhum/v2.x/tutorials/test-doubles/mocks)

- PR: https://github.com/drashland/website-v2/pull/193

### CODE REVIEW

File: console/build_esm_lib.ts [hunk 1]
```diff
@@ -83,11 +128,62 @@ function removeTsExtensions(filename: string): void {
...
+}
+
+/**
+ * Is the given option enabled?
+ * @param option The name of the option (e.g., `--debug`).
+ * @returns True if yes, false if no.
+ */
+function optionEnabled(option: string): boolean {
+  const optionIndex = args.indexOf(option);
+  const enabled = optionIndex != -1;
```
Comment (ebebbington) [inline console/build_esm_lib.ts:161]:
```suggestion
  const enabled = optionIndex !== -1;
```

File: console/build_esm_lib.ts [hunk 2]
```diff
@@ -83,11 +128,62 @@ function removeTsExtensions(filename: string): void {
...
+ * What is the given option value?
+ * @param option The name of the option (e.g., `--workspace`).
+ * @returns The value of the option if the option exists.
+ */
+function optionValue(option: string): boolean {
+  const extractedOption = args.filter((arg: string) => {
+    if (arg.includes(option)) {
+      return true;
+    }
+  });
```
Comment (ebebbington) [inline console/build_esm_lib.ts:180]:
```suggestion
  const extractedOption = args.filter(arg => arg.includes(option));
```

**Review by ebebbington — CHANGES_REQUESTED** [review]
just couple minor nitpicky things

**Review by Guergeiro — APPROVED** [review]

**Comment by crookse** [comment]
:heart: @Guergeiro @ebebbington 

I'm writing the docs now

**Review by Guergeiro — APPROVED** [review]

**Comment by crookse** [comment]
@Guergeiro @ebebbington , docs are up: https://github.com/drashland/website-v2/pull/193

**Pull request merged**